File: Pommora/Pommora/AtomicIO/SchemaTransaction.swift
import Foundation

/// Errors thrown by `SchemaTransaction`. Surface to managers via the existing
/// pendingError toast pipeline (per Pommora discipline).
enum SchemaTransactionError: Error, Sendable {
    /// A staged temp-file write failed during phase 1 (stage). The first URL is the
    /// target whose temp write blew up; the wrapped error is the underlying cause.
    case stageFailed(target: URL, underlying: any Error)
    /// A rename from temp → target failed during phase 2 (commit). Already-renamed
    /// targets have been rolled back to their backups; remaining temps were deleted.
    case commitFailed(target: URL, underlying: any Error)
}

/// Atomic multi-file commit primitive for schema mutations that touch >1 file.
///
/// Used by schema mutations that touch >1 file: schema type-change with
/// value-drop (writes schema sidecar + every affected member file), ID-rekey
/// migration (writes schema + every member file per Type), and move-strip
/// (writes source + destination member files).
///
/// ## Pattern (two-phase commit)
///
/// 1. **Stage:** every `stage(payload:to:)` records a (payload, target, temp)
///    triple. `commit()` writes all payloads to sibling `<target>.txn-<ulid>`
///    temp files. If ANY stage write fails, every already-written temp is
///    deleted and `SchemaTransactionError.stageFailed` is thrown.
/// 2. **Commit:** every staged temp is moved over its target via
///    `FileManager.moveItem`. Pre-existing targets are first moved aside to a
///    sibling `<target>.bak-<ulid>` backup. If ANY rename fails, the backups
///    are restored over their targets (best-effort) and remaining temps are
///    deleted; `SchemaTransactionError.commitFailed` is thrown. On success,
///    backups are deleted.
///
/// ## Idempotence
///
/// Before staging, every target's parent directory is swept for stale
/// `*.txn-*` and `*.bak-*` siblings (likely left over from a previous
/// crashed commit). They are deleted unconditionally — they have no value
/// across processes.
///
/// ## Single-file commits
///
/// Use `AtomicJSON.write` directly for single-file commits. `SchemaTransaction`
/// is overhead-only when the mutation is one file.
final class SchemaTransaction {

    private struct StagedWrite {
        let payload: Data
        let target: URL
        let temp: URL
    }

    private var pending: [StagedWrite] = []

    init() {}

    /// Stage a payload to be written to `url` on the next `commit()`. Multiple
    /// stages to the same target overwrite each other (last-stage-wins) — not
    /// generally useful but defined.
    func stage(payload: Data, to url: URL) {
        let temp = url.appendingPathExtension("txn-\(ULID.generate())")
        pending.append(StagedWrite(payload: payload, target: url, temp: temp))
    }

    /// Convenience: encode + stage in one call for any `Codable` value using
    /// the same JSON encoding shape as `AtomicJSON.write`.
    func stage<T: Codable>(_ value: T, to url: URL) throws {
        stage(payload: try AtomicJSON.encode(value), to: url)
    }

    /// Two-phase commit. Throws if any stage or rename fails; on throw, the
    /// filesystem is restored to its pre-commit state (best-effort).
    func commit() throws {
        cleanStaleTemps()

        // Phase 1: stage all temps.
        var stagedTemps: [URL] = []
        for write in pending {
            do {
                try write.payload.write(to: write.temp, options: [.atomic])
                stagedTemps.append(write.temp)
            } catch {
                // Roll back already-staged temps.
                for url in stagedTemps {
                    try? FileManager.default.removeItem(at: url)
                }
                throw SchemaTransactionError.stageFailed(target: write.target, underlying: error)
            }
        }

        // Phase 2: rename all temps to targets in deterministic order.
        var renamed: [(target: URL, backup: URL?)] = []
        for write in pending {
            do {
                var backup: URL?
                if FileManager.default.fileExists(atPath: write.target.path) {
                    let backupURL = write.target.appendingPathExtension("bak-\(ULID.generate())")
                    try FileManager.default.moveItem(at: write.target, to: backupURL)
                    backup = backupURL
                }
                try FileManager.default.moveItem(at: write.temp, to: write.target)
                renamed.append((write.target, backup))
            } catch {
                // Rollback: restore backups (reverse order so nested updates restore correctly),
                // delete any remaining temps.
                for (target, backup) in renamed.reversed() {
                    try? FileManager.default.removeItem(at: target)
                    if let backup {
                        try? FileManager.default.moveItem(at: backup, to: target)
                    }
                }
                // Also restore the backup we'd just moved aside (before the move failure),
                // since `renamed` doesn't include this iteration.
                if FileManager.default.fileExists(atPath: write.temp.path) {
                    try? FileManager.default.removeItem(at: write.temp)
                }
                // Any further pending temps haven't been touched yet; they get cleaned next call.
                throw SchemaTransactionError.commitFailed(target: write.target, underlying: error)
            }
        }

        // Phase 3: success — delete backups.
        for (_, backup) in renamed {
            if let backup {
                try? FileManager.default.removeItem(at: backup)
            }
        }

        // Clear pending so a transaction object can be reused (not common, but
        // doesn't cost anything).
        pending.removeAll()
    }

    /// Sweep every staged target's parent directory for `*.txn-*` and `*.bak-*`
    /// siblings left over from a previous crashed commit and delete them.
    private func cleanStaleTemps() {
        let parentDirs = Set(pending.map { $0.target.deletingLastPathComponent() })
        let txnMarker = ".txn-"
        let bakMarker = ".bak-"
        for dir in parentDirs {
            guard
                let contents = try? FileManager.default.contentsOfDirectory(
                    at: dir, includingPropertiesForKeys: nil)
            else { continue }
            for entry in contents {
                let name = entry.lastPathComponent
                if name.contains(txnMarker) || name.contains(bakMarker) {
                    try? FileManager.default.removeItem(at: entry)
                }
            }
        }
    }
}

File: React/src/main/io/schemaTransaction.ts
// Atomic multi-file commit for schema mutations that touch more than one file. A
// delete-property or a lossy type-change rewrites the type sidecar AND strips the
// property from every member page — those writes must land together-or-not-at-all. A
// filesystem has no real multi-file transaction, so this fakes it with a two-phase
// commit: stage every payload to a sibling temp, then rename each over its target
// (backing up any existing target); any failure rolls the filesystem back. A crash
// mid-commit is self-healed by the stale-temp sweep on the next commit. Single-file
// writes should use atomicWriteFile directly — this is overhead-only for one file.
// Mirrors Swift's SchemaTransaction.

import { writeFile, rename, readdir, unlink } from 'node:fs/promises'
import { join, dirname } from 'node:path'
import { newId } from '../ids'
import { pathExists } from './atomicWrite'

interface Staged {
  target: string
  content: string
  temp: string
}

export class SchemaTransaction {
  private pending: Staged[] = []

  /** Stage a UTF-8 payload to write to `target` on the next `commit()`. Restaging the
   *  same target replaces the prior stage (last-stage-wins). */
  stage(target: string, content: string): void {
    const staged: Staged = { target, content, temp: `${target}.txn-${newId()}` }
    const at = this.pending.findIndex((p) => p.target === target)
    if (at >= 0) this.pending[at] = staged
    else this.pending.push(staged)
  }

  /** How many distinct files are staged. */
  get size(): number {
    return this.pending.length
  }

  /** Two-phase commit. On any failure the filesystem is restored to its pre-commit
   *  state (best-effort) and the error is rethrown. */
  async commit(): Promise<void> {
    await this.cleanStaleTemps()

    // Phase 1: write every payload to its temp sibling.
    const written: string[] = []
    for (const w of this.pending) {
      try {
        await writeFile(w.temp, w.content, 'utf8')
        written.push(w.temp)
      } catch (e) {
        for (const t of written) await unlink(t).catch(() => {})
        throw new Error(`SchemaTransaction stage failed for ${w.target}: ${String(e)}`)
      }
    }

    // Phase 2: rename each temp over its target, backing up any existing target.
    const renamed: { target: string; backup: string | null }[] = []
    for (const w of this.pending) {
      // `backup` is hoisted so the catch can restore THIS entry too: if the target was
      // already moved aside but the temp→target rename then failed, the original lives
      // only in `backup` and isn't in `renamed` yet — without this it would be lost.
      let backup: string | null = null
      try {
        if (await pathExists(w.target)) {
          backup = `${w.target}.bak-${newId()}`
          await rename(w.target, backup)
        }
        await rename(w.temp, w.target)
        renamed.push({ target: w.target, backup })
      } catch (e) {
        // Roll back: restore THIS entry's backup first, then the already-renamed ones
        // (reverse order), then delete any remaining temps.
        if (backup) await rename(backup, w.target).catch(() => {})
        for (const r of [...renamed].reverse()) {
          await unlink(r.target).catch(() => {})
          if (r.backup) await rename(r.backup, r.target).catch(() => {})
        }
        for (const t of this.pending) await unlink(t.temp).catch(() => {})
        throw new Error(`SchemaTransaction commit failed for ${w.target}: ${String(e)}`)
      }
    }

    // Phase 3: success — delete backups; clear pending (the object is reusable).
    for (const r of renamed) if (r.backup) await unlink(r.backup).catch(() => {})
    this.pending = []
  }

  /** Sweep every staged target's parent dir for stale `*.txn-*` temps left by a previous
   *  crashed commit (a temp is uncommitted, so it's always safe to delete). `*.bak-*` files
   *  are deliberately NOT swept: a leftover backup is the ORIGINAL content of a target whose
   *  commit crashed after the move-aside but before the temp landed — it may be the only
   *  copy, and a concurrent transaction's live backups would also be hit. Successful commits
   *  delete their own backups in phase 3; a crashed one's backup is left for recovery. */
  private async cleanStaleTemps(): Promise<void> {
    const dirs = new Set(this.pending.map((p) => dirname(p.target)))
    for (const dir of dirs) {
      let entries: string[]
      try {
        entries = await readdir(dir)
      } catch {
        continue
      }
      for (const name of entries) {
        if (name.includes('.txn-')) await unlink(join(dir, name)).catch(() => {})
      }
    }
  }
}
